### PR TITLE
[Compaction] Fix the delete reference timestamp differences between blob store compactor and stats

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Will add a test to make sure the stats and compactor returns the same data size

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java [hunk 1]
```diff
@@ -1286,15 +1286,7 @@ private boolean isDeleteTombstoneRemovable(IndexEntry deleteIndexEntry, IndexSeg
       // TTL update + DELETE -> DELETE
       // PUT + TTL update -> PUT (the one relevant to this comment)
       // PUT + TTL update + DELETE -> DELETE
-      // TODO: move this blob store stats
-      Offset startOffsetOfLastIndexSegmentForDeleteCheck = getStartOffsetOfLastIndexSegmentForDeleteCheck();
```
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java:1290]: Add a to-do that we can remove `getStartOffsetOfLastIndexSegmentForDeleteCheck()` method once we verified the correctness.

**Review by jsjtzyy — COMMENTED** [review]
Mostly looks good to me. Will approve once comments are addressed.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java [hunk 2]
```diff
@@ -285,7 +310,7 @@ long getMaxBlobSize() {
    * @throws StoreException
    */
   Pair<Long, NavigableMap<LogSegmentName, Long>> getValidDataSizeByLogSegment(TimeRange timeRange,
-      long expiryReferenceTime) throws StoreException {
+      long expiryReferenceTime, FileSpan fileSpanUnderCompaction) throws StoreException {
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:313]: Removing it in this PR as well.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java [hunk 3]
```diff
@@ -689,24 +720,33 @@ private void forEachValidIndexEntry(IndexSegment indexSegment, long deleteRefere
...
       return false;
     }
+    // Validity of ttl update is determined by the logic like this
+    // Is there PUT before this ttl update
+    //   NO: invalid
+    //   YES: is the final state a delete and out of retention
+    //     NO: valid
+    //     YES: is ttl update NOT in the same log as put || is put NOT under compaction
+    //       YES: valid
+    //       NO: invalid
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:735]: TTL_UPDATE is a tricky index value here, although I don't agree with the previous code, but I have to follow it. 
The TTL_UPDATE is invalid only when the PUT is going to be compacted. So if the PUT is not going to be compacted, then TTL_UPDATE is valid. 

So for your case, final state is delete (out of retention), PUT and TTL_UPDATE in the same log, but PUT is not within compaction file span, then PUT is not under compacted. So TTL_UPDATE can be valid. 

Can we just invalidate TTL_UPDATE in this case? Yeah, I would agree so. But the previous code is structure in this way. I might as well just follow that.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java [hunk 4]
```diff
@@ -564,20 +591,24 @@ private void enqueueNewValue(StoreKey key, IndexValue newValue, IndexValue origi
...
+   * @param expiryReferenceTimeInMs the reference time in ms until which expiration are relevant
    */
   private void removeDeleteTombStonesFromValidSize(Collection<IndexFinalState> indexFinalStates,
-      NavigableMap<LogSegmentName, Long> validSizePerLogSegment) {
+      NavigableMap<LogSegmentName, Long> validSizePerLogSegment, long expiryReferenceTimeInMs) {
     for (IndexFinalState finalState : indexFinalStates) {
       if (finalState.isDelete()) {
-        if (finalState.getExpirationTime() != Utils.Infinite_Time) {
+        if (finalState.getExpirationTime() != Utils.Infinite_Time && isExpired(finalState.getExpirationTime(),
+            expiryReferenceTimeInMs)) {
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java:611]: that's right, I want to make sure we are making the stats returned by BlobStoreStats as close to the real value as possible.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java:1290]: Will remove it.

**Review by jsjtzyy — APPROVED** [review]

**Comment by zzmao** [comment]
LGTM! The code looks much cleaner now.

**Review by zzmao — APPROVED** [review]

**Pull request merged**